diff --git a/plume_d3d12.cpp b/plume_d3d12.cpp
index 2cdcf18..172a434 100644
--- a/plume_d3d12.cpp
+++ b/plume_d3d12.cpp
@@ -2312,7 +2312,10 @@ namespace plume {
 
         const D3D12_TEXTURE_COPY_LOCATION copyDstLocation = toD3D12(dstLocation);
         const D3D12_TEXTURE_COPY_LOCATION copySrcLocation = toD3D12(srcLocation);
-        setSamplePositions(dstLocation.texture);
+        setSamplePositions(
+            (dstLocation.texture != nullptr)
+                ? dstLocation.texture
+                : srcLocation.texture);
         d3d->CopyTextureRegion(&copyDstLocation, dstX, dstY, dstZ, &copySrcLocation, (srcBox != nullptr) ? &copyBox : nullptr);
         resetSamplePositions();
     }
